### PR TITLE
Adds support for Stylers and Formatters in Tables

### PR DESCRIPTION
**Pull request by adamgalway-r7**
Allows users to pass in classes that can be used to format and style all values in a column. Both Formatters and Stylers work by passing the contents of a field into the classes' `format` and `style` method respectively. 

Formatters are called when a row is added to a table, and result in the width of a column being increased. They should be used for changes that change the content of a field IE Date/Time Formatters.

Stylers are called when `to_s` is called for any Rex::Text::Table, and their alterations do not result in an increase in a column's width. They should be used for changes that only affect the appearance of a field  without affecting it's length IE Colorization.

[Relevant PR](https://github.com/rapid7/metasploit-framework/pull/13039)

### CODE REVIEW

File: spec/rex/text/table_spec.rb [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      ======
+
+      IHAVEBEENSTYLEDColumn 1  IHAVEBEENSTYLEDColumn 2                                            IHAVEBEENSTYLEDColumn 3
+      AAAAA     BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB  CCCCCCCCCCCCCCC
+    TABLE
+  end
+
+end
+
+class Formatter
```
Comment (adfoster-r7) [inline spec/rex/text/table_spec.rb]: Just a heads up, defining classes here can lead to all sorts of unexpected behavior/issues due to these classes now being defined globally - https://makandracards.com/makandra/47189-rspec-how-to-define-classes-for-specs

A solution is posted in the above link, or you can maybe do:

```ruby
  let(:formatter) do
    Formatter = Class.new do
      def format(str)
        "IHAVEBEENFORMATTED#{str}"
     end
    end

    Formatter.new
  end

  let(:styler) do
    Styler = Class.new do
      def style(str)
        "IHAVEBEENSTYLED#{str}" 
      end
    end

    Styler.new
  end
```

Note the name `styler` now returns an instance; We can update the use of it in our tests now too:

```diff
- stylers: [Styler.new]
+ stylers: [styler]
```

File: spec/rex/text/table_spec.rb [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    options = {
+      'Header'     => 'Header', 
+      'SearchTerm' => ['jim','bob'],
+      'Columns' => [
+        'Column 1',
+        {
+          column: 'Column 2',
+          stylers: [Styler.new]
+        },
```
Comment (adfoster-r7) [inline spec/rex/text/table_spec.rb]: I prefer the symbol syntax here, but on reflection, what are your thoughts of being consistent with the existing string API?

```suggestion
        {
          'Column' => 'Column 2',
          'Stylers' => [Styler.new]
        },
```

File: lib/rex/text/table.rb [hunk 3]
```diff
@@ -355,17 +368,15 @@ def columns_to_s # :nodoc:
         nameline << padding
         remainder = padding.length - cellpad
         remainder = 0 if remainder < 0
-        barline << (' ' * (cellpad + remainder))
```
Comment (adfoster-r7) [inline lib/rex/text/table.rb]: Hm - just confirming, does this break backwards compatibility in any way? We'd want all existing tables to be formatted visually in the same way

File: lib/rex/text/table.rb [hunk 4]
```diff
@@ -425,6 +436,69 @@ def pad(chr, buf, colidx, use_cell_pad = true) # :nodoc:
...
+
+    self.formatters[col_name].each do |f|
+      str_cp = f.format(str_cp)    
+    end
+
+    str_cp
+  end
+  
+  def style_table_field(str, col_name)
+    if col_name.nil? || !self.stylers.has_key?(col_name)
```
Comment (adfoster-r7) [inline lib/rex/text/table.rb]:
```suggestion
    unless self.stylers.has_key?(col_name)
```

File: lib/rex/text/table.rb [hunk 5]
```diff
@@ -425,6 +436,69 @@ def pad(chr, buf, colidx, use_cell_pad = true) # :nodoc:
     return val
   end
 
+  def format_table_field(str, col_name)
+    if col_name.nil? || !self.formatters.has_key?(col_name)
```
Comment (adfoster-r7) [inline lib/rex/text/table.rb]:
```suggestion
    unless self.formatters.has_key?(col_name)
```

File: spec/rex/text/table_spec.rb (hunk 2 above)
Comment (adamgalway-r7) [reply spec/rex/text/table_spec.rb]: I like my efficiencies and I'd definitely prefer to use symbols for the performance improvement, but I doubt it's worth changing everything in the API and using a half & half approach is even worse. So I'd say keep everything consistent with the existing api and use hash rockets

**Pull request merged**